core/dimensions.cpp

```diff
@@ -218,4 +229,65 @@ Dimensions transpose(const Dimensions &dims, std::vector<Dim> labels) {
   return {labels, shape};
 }
 
+/// Fold one dim into multiple dims
+///
+/// Go through the old dims and:
+/// - if the dim does not equal the dim that is being stacked, copy dim/shape
+/// - if the dim equals the dim to be stacked, replace by stack of new dims
+///
+/// Note that addInner will protect against inserting new dims that already
+/// exist in the old dims.
+/// If from_dim is not found in old_dims, the new dims are identical to the
+/// old_dims (this occurs when folding a DataArray whose coordinates do not all
+/// necessarily contain from_dim).
+Dimensions fold(const Dimensions &old_dims, const Dim from_dim,
+                const Dimensions &to_dims) {
+  Dimensions new_dims;
+  for (const auto dim : old_dims.labels())
+    if (dim != from_dim)
+      new_dims.addInner(dim, old_dims[dim]);
+    else
+      for (const auto lab : to_dims.labels())
+        new_dims.addInner(lab, to_dims[lab]);
+  return new_dims;
+}
+
+/// Flatten multiple dims into one
+///
+/// Go through the old dims and:
+/// - if the dim is contained in the list of dims to be flattened, add the new
+///   dim once
+/// - if not, copy the dim/shape
+Dimensions flatten(const Dimensions &old_dims,
+                   const scipp::span<const Dim> from_labels, const Dim to_dim) {
+  Dimensions from_dims;
+  for (const auto dim : from_labels)
+    if (old_dims.contains(dim))
+      from_dims.addInner(dim, old_dims[dim]);
+
+  // Only allow reshaping contiguous dimensions.
+  // Note that isContiguousIn only allows for inner contiguous blocks,
+  // and contains(dimensions) ignores dimension order.
+  Dimensions intersect = intersection(old_dims, from_dims);
+  for (scipp::index i = 0; i < intersect.ndim() - 1; ++i)
+    if (old_dims.index(intersect.label(i + 1)) !=
+            old_dims.index(intersect.label(i)) + 1 ||
+        from_dims.index(intersect.label(i + 1)) !=
+            from_dims.index(intersect.label(i)) + 1)
+      throw except::DimensionError(
+          "Flatten: can only flatten a contiguous set of dimensions. "
+          "The order of the dimensions to flatten must also match the order of "
+          "the old dimensions.");
+
+  Dimensions new_dims;
+  for (const auto dim : old_dims.labels())
+    if (from_dims.contains(dim)) {
+      if (!new_dims.contains(to_dim))
+        new_dims.addInner(to_dim, from_dims.volume());
+    } else {
+      new_dims.addInner(dim, old_dims[dim]);
+    }
+  return new_dims;
```

Comment:
Having trouble understanding this code, can we write it simpler?

```python
Dimensions flatten(const Dimensions &old_dims,
                   const scipp::span<const Dim> from_labels, const Dim to_dim) {
  const auto i = old_dims.index(from_labels.front());
  if (!std::equal(from_labels.begin(), from_labels.end(), old_dims.labels().subspan(i).begin())
    throw ...;
  Dimensions new_dims;
  for (int64_t j=0; j<old_dims.ndim(); ++j) {
    const auto dim = old_dims.label(j);
    const auto size = old_dims.size(j);
    if (j<i || j>=i+scipp::size(from_labels))
      new_dims.addInner(dim, size);
    else if (j == i)
      new_dims.addInner(to_dim, size);
    else
      new_dims.resize(to_dim, size * new_dims.size(to_dim));
  }
  return new_dims;
}
```
or something like that? And remove `intersection`?

Reply:
The `from_labels` are not necessarily contained in the `old_dims`. This happens if, e.g. the you have a 3d array with dims `x,y,z` and you are flattening `x,y` but you have a 1d coordinate along `z`.
None of the `from_labels` are found in the `z` coord, because the `from_labels` are the ones passed to the function, rather than taking out only the labels relevant to the current variable.
(It was easier to do things like this because when requiring broadcast it was easier to supply all the `from_labels`)

The code above doesn't throw in this case, but i think your
`if (!std::equal(from_labels.begin(), from_labels.end(), old_dims.labels().subspan(i).begin())`
would throw?

That's why i need the intersection: if i try to flatten `[x, y]` on a `z` coord, it just gives an empty set of dims and nothing happens.

Reply:
Then my code above, but as a first step remove anything from `from_labels` not found in `old_dims`?

Reply:
But I would then argue that it is not much cleaner than what is already there?

I also find the
```C++
    if (j<i || j>=i+scipp::size(from_labels))
      new_dims.addInner(dim, size);
    else if (j == i)
      new_dims.addInner(to_dim, size);
    else
      new_dims.resize(to_dim, size * new_dims.size(to_dim));
```
part not so easy to follow.

In addition, `intersection` may be useful elsewhere in the future?
What I liked about working with the `Dimensions` class is that it was easy to understand what all the member functions did just from their names, and virtually every time I needed something (like `contains` or `merge`) it was already there (apart from the `intersection`).
I think it's a valid thing to add, although I agree that strictly speaking, it probably should be needed in more than one place before adding it as a separate function.

However, a counter argument to that is that if I implement an intersect in a not so obvious way inside the body of the function, if someone else later is looking for an intersect method in the class, they will probably not find it because it is 'hidden', and may implement it all over again.

And removing anything from `from_labels` not found in `old_dims` is basically an intersect, isn't it?

I agree that the check for contiguous dimensions is not super pretty, and I will try to make it cleaner with the `std::equal` and `subspan()`.

Reply:
Sure, `intersection` may be useful, but it is quite ambiguous as to what it should do (use order of first, of second, or throw?). Right now the implementation of `intersection` uses a certain definition and `flatten` seems to (silently?) rely on that? This makes things confusing.